confluent_kafka/schema_registry/schema_registry_client.py

```diff
@@ -0,0 +1,588 @@
+
+            response = self.session.request(
+                method, url="/".join([self.base_url, url]),
+                headers=_headers, data=body)
+
+            try:
+                if 200 <= response.status_code <= 299:
+                    return response.json(cls=decoder, **kwargs)
+                raise SchemaRegistryClientError(response.status_code,
+                                                response.json().get('message'))
+            except ValueError:
+                raise SchemaRegistryClientError(response.status_code,
+                                                response.content)
+
+    def register_schema(self, subject_name, schema):
+        """
+        POST /subjects/(string: subject)/versions
+
+        Registers a schema under ``subject_name``.
+        Results are cached in the ``schema`` so subsequent requests will not
+        require another trip the Schema Registry.
+
+        Args:
+            - subject_name (str): subject to register a schema under
+            - schema (Schema): Schema (in JSON string format) to register.
+
+        Returns:
+            Version: Registration information
+
+        Raises:
+            SchemaRegistryClientError: if Schema violates this subject's
+            Compatibility policy or is otherwise invalid.
+
+        .. _Schema Registry API Reference:
+            https://docs.confluent.io/current/schema-registry/develop/api.html#post--subjects-(string-%20subject)-versions
+
+        """  # noqa: E501
+        version = schema.subjects.get(subject_name, None)
+        if version:
+            return version
+
+        self._rest_client.POST(
+            url_formatter(VERSIONS, subject_name),
+            body=schema, encoder=_SchemaJSONEncoder)
+
+        # get registration information
+        return self._rest_client.POST(
+            url_formatter(SUBJECT, subject_name),
+            body=schema, encoder=_SchemaJSONEncoder, decoder=_VersionJSONDecoder,
+            schema=schema)
+
+    def get_schema(self, schema_id):
+        """
+        GET /schemas/ids/{int: schema_id}
+
+        Fetches the schema associated with ``schema_id`` from the
+        Schema Registry. The result is cached so subsequent attempts will not
+        require an additional trip to the Schema Registry.
+
+        Args:
+            - schema_id (int): Schema id
+
+        Returns:
+            Schema: Schema string identified by the ``schema_id``
+
+        Raises:
+            SchemaRegistryClientError: If schema can't be found
+
+        .. _Schema Registry API Reference:
+            https://docs.confluent.io/current/schema-registry/develop/api.html#get--schemas-ids-int-%20id
+
+        """  # noqa: E501
+        version = self.id_to_schema.get(schema_id, None)
+
+        if version is not None:
+            return version
+
+        url = url_formatter(SCHEMAS, schema_id)
+        schema = self._rest_client.GET(url, decoder=_SchemaJSONDecoder)
```

Comment:
if you're producing messages in a web request handler (common use case), this will often be configured for multiple threads, so it would be good if this was thread safe.